src/efftester.ml

```diff
@@ -99,14 +103,25 @@ type lit =
   | LitStr of string
   | LitList of lit list
 
+(** type term is used to represent all syntax constructs (of OCaml) available in our
+generator  *)
 type term =
   | Lit of lit
   | Variable of etype * variable
+  (* [Constructor (type, name, load)] is used to construct ADT variants *)
+  | Constructor of etype * string * term option
+  (* [PatternMatch typ matched_trm branches eff]
+    Note: in type (term * term) list, the first part of the tuple must be a value (no evaluation)
+  *)
+  | PatternMatch of etype * term * (term * term) list * eff
```

Comment:
In `(term * term) list`, I think the `term` on the left should be a new syntactic category dedicated to patterns, I guess just `pattern`. There are various choices that we could support, from the simplest

```ocaml
type pattern =
  | Constructor of etype * constructor_name * variable list
```

which forces patterns to start a constructor and have only variables as arguments, or something more complex like

```ocaml
type pattern =
  | Variable of variable
  | Constructor of etype * constructor_name * pattern list
```

Reply:
Do we use pattern *list* in case our pattern is something like this `x::xs`? Are there other reasons for using a list? Sorry it's evening already in my country.

Reply:
There are no stupid questions! Yes, `(::)` is a constructor that expects two arguments, so it would expect a list (of size 2) in both the expression and the pattern case.

Reply:
Looking at my implementation of 
```ocaml
type pattern =
  | PattVar of variable
  | PattConstr of etype * constructor_name * pattern list
```
I understand that this type's semantics is unspecified. What I mean is, what does `PattConstr (Option Int, "Some", [PattConstr (Option Int, "None", []); PattVar "x"])` map to? 

In my implementation, I haven't considered a case with more than one element in the list (e.g. Some (Some x)), which means it misses out on cases.

Reply:
As we do not have tuples yet, should I simply have cases for an empty and single-element pattern case? I am concerned about `term_to_ocaml` here.

Reply:
If you introduce a new syntactic category of `pattern` you should extend `term_to_ocaml` with a helper `pattern_to_ocaml` to convert patterns. It is very much type-directed programming 😀

For inspiration, have a look at the OCaml manual's grammar for the syntactic categories:
http://caml.inria.fr/pub/docs/manual-ocaml/patterns.html
http://caml.inria.fr/pub/docs/manual-ocaml/expr.html
Notice how `match` in expressions uses a category of "pattern-matching" which uses something like your `pattern`.